core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -2380,10 +2381,14 @@ class SparkContext(config: SparkConf) extends Logging {
 
   /** Reports heartbeat metrics for the driver. */
   private def reportHeartBeat(): Unit = {
-    val driverUpdates = _heartbeater.getCurrentMetrics()
```

Comment:
Would it be useful to poll more frequently for driver metrics as well?

Reply:
My thinking, as I commented in the JIRA, was that most memory issues are with executors rather than with the driver. So I thought it was much less important to poll frequently in the driver.
That said, we can poll more frequently in the driver as well and keep track of peaks since the last heartbeat. The only issue though is that in the driver we don't know what stages are active each time we poll; we rely on the EventLoggingListener, which knows when a stage has started or ended, to attribute the driver's update to the stage. Peaks since the last heartbeat could cross stage boundaries. But maybe we don't care?

Reply:
We've seen some issues with OOM on the driver, but it is much more common for executors, so agreed that it is not as important for the driver. For accurate stage information, one option could be to pass in a pointer to the DAGScheduler, and add a method to DAGScheduler to return the running stages.

Thanks for your work improving the accuracy of the metrics. Have a great vacation! Please let me know if you have any questions about the comments, or want to discuss anything when you get back.

Reply:
I'd punt on more frequent polling in the driver for this change. Let's open another JIRA for that if someone would like to take it up.